File: server/nudge_policy.py
# Ported from gauge@2157433 server/nudge_policy.py; adapted per docs/plans/2026-08-15-phase1-one-repo-one-engine.md
# Moved from server/watch/nudge_policy.py (Foundation A, 2026-08-24) so the
# phone's realtime path (server/audio_pipeline.py, later track) can share the
# SAME escalation brain as the watch instead of growing its own. The old
# module path is kept as a thin re-export — every existing import still works.
"""Shared nudge/escalation policy: vector events -> per-channel haptic levels.

Mirror contract: this is the server-side counterpart of the on-watch
NudgeStateMachine at
apps/watch/shared/src/commonMain/kotlin/app/gauge/shared/NudgeStateMachine.kt
— thresholds/semantics (per-channel hysteresis, cooldown, half-up rounding)
must stay identical between the two. The cross-language golden vectors in
server/tests/fixtures/policy_vectors/nudge_policy.json (driven here by
server/tests/test_nudge_policy_vectors.py) are the executable form of that
contract; a Kotlin/TypeScript port should consume the same file.

Channel model: a "channel" is an independent escalation lane with its own
level and its own cooldown clock (the watch uses "A" = voice vectors, "B" =
heart-rate). The policy is parameterized by which channels it runs — the
default ``("A", "B")`` is exactly the shipped watch behaviour; a second
caller that only has one lane (e.g. the phone's realtime coaching path) can
run ``channels=("A",)`` and gets identical single-channel semantics without
a phantom idle "B" in its ``current()`` output. Channel *labels* are still
constrained to ``watch.models.Channel`` because ``NudgeEvent.channel`` is
typed by it — a new lane means a new Literal member there, not an ad-hoc
string here.
"""

from __future__ import annotations

import os
from typing import Sequence

from watch.models import Channel, NudgeEvent, VectorEvent, VectorSubscription

# The shipped watch's two lanes, in emission order. Emission order matters
# for byte-for-byte compatibility: ``on_events`` returns at most one nudge per
# channel, in THIS order, and server/tests/watch/test_nudge_policy.py's
# assertions (and the watch's WS consumer) rely on A before B.
DEFAULT_CHANNELS: tuple[Channel, ...] = ("A", "B")

# --- hold-N hysteresis on the loudness ladder (2026-09-20) -------------------
# The ladder used to escalate off a SINGLE 1 s window that cleared +6 dB over
# the wearer's baseline. Measured over 32 h of real conversation (AMI + SBCSAE,
# scripts/heat_map.py) that instantaneous crossing is most of the dose: the
# median AMI meeting delivered 97 buzzes/hour and the median SBCSAE recording
# 80 — in conversations where nobody is angry. Requiring the first rung to HOLD
# for three consecutive windows before the ladder may climb halves it (46.4 and
# 36.0 respectively) without any new signal, because a single loud window is a
# laugh, a cough or a door, and three in a row is a raised voice.
#
# Reference implementation and the measurement: ``sustained()`` in
# scripts/conversation_audit.py. ``LoudnessHold`` below reproduces it exactly
# when fed one 1 s window per call.
#
# The window the watch/server PCM paths run on: SentinelController's
# MAX_SLICE_BYTES 32000 / 2 bytes / 16 kHz.
WINDOW_S = 1.0
#: Seconds the first rung must hold before ANY upward escalation on the
#: loudness lane. 0 or 1 reproduces the pre-2026-09-20 behaviour exactly (one
#: qualifying window is enough), which is how the old golden vectors are pinned.
HEAT_HOLD_S_DEFAULT = 3.0
#: Server-side override, read once per policy construction.
HEAT_HOLD_ENV = "MINDSHIFT_HEAT_HOLD_S"
#: The vector the hold gates. Only the LOUDNESS lane: ``aggressive_tone`` is
#: about the words, is a better signal, and rides the same channel untouched.
HOLD_VECTOR = "yelling"


def heat_hold_s() -> float:
    """Configured hold, in seconds. ``MINDSHIFT_HEAT_HOLD_S`` overrides
    [HEAT_HOLD_S_DEFAULT]; an unparseable or negative value falls back to the
    default rather than disabling the gate silently."""
    raw = (os.getenv(HEAT_HOLD_ENV) or "").strip()
    if not raw:
        return HEAT_HOLD_S_DEFAULT
    try:
        value = float(raw)
    except ValueError:
        return HEAT_HOLD_S_DEFAULT
    return value if value >= 0.0 else HEAT_HOLD_S_DEFAULT


class LoudnessHold:
    """Has the ladder's first rung held long enough to escalate?

    Exactly ``scripts/conversation_audit.py``'s ``sustained()`` when fed one
    1 s window per call: a run of consecutive qualifying observations, reset by
    the first that does not qualify, and the gate opens once the run reaches
    ``hold_s``.

    Turn-driven callers (the phone's fast loop, ``watch/relay.py``) observe once
    per TURN rather than once per window, so an observation carries how many
    seconds of audio it covers — a 4 s turn that read as loud is four windows of
    hold, not one. Every observation is worth at least one window, which is what
    makes ``hold_s <= 1`` byte-identical to the pre-hold behaviour on every path.
    """

    def __init__(self, hold_s: float | None = None, window_s: float = WINDOW_S):
        self.hold_s = heat_hold_s() if hold_s is None else float(hold_s)
        self.window_s = window_s
        self._run_s = 0.0

    @property
    def run_s(self) -> float:
        """Seconds of consecutive qualifying observation so far."""
        return self._run_s

    def _next_run(self, loud: bool, observed_s: float) -> float:
        if not loud:
            return 0.0
        return self._run_s + max(observed_s, self.window_s)

    def peek(self, loud: bool, observed_s: float = WINDOW_S) -> bool:
        """Would this observation open the gate? Pure — the phone's instant
        haptic tier asks before the same turn's policy tick advances it."""
        return self._next_run(loud, observed_s) >= self.hold_s

    def observe(self, loud: bool, observed_s: float = WINDOW_S) -> bool:
        """Advance the run and report whether this observation may escalate."""
        self._run_s = self._next_run(loud, observed_s)
        return self._run_s >= self.hold_s


class NudgePolicy:
    """Transforms vector events into per-channel haptic nudges with hysteresis.

    Channel level = max over subscribed+haptics-on vectors of round(level × sensitivity) clamped 0–3.
    Emits NudgeEvent only when a channel's level changes.
    De-escalates (drops one level) after cooldown_s seconds with no event ≥ current level.
    Sustained events at current level refresh the decay clock without emitting.

    Hold-N hysteresis (2026-09-20): a ``yelling`` event may only raise a channel
    once the first rung has held for ``hold_s`` seconds — see [LoudnessHold].
    A gated-out loudness observation counts as "not loud" for THIS call, so it
    neither escalates nor refreshes the decay clock, exactly as the measured
    replay models it (``conversation_audit.replay``'s ``gate``).
    """

    def __init__(
        self,
        subs: list[VectorSubscription],
        cooldown_s: float = 20.0,
        channels: Sequence[Channel] = DEFAULT_CHANNELS,
        hold_s: float | None = None,
    ):
        if not channels:
            raise ValueError("NudgePolicy needs at least one channel")
        self.subs = subs
        self.cooldown_s = cooldown_s
        #: The loudness lane's hysteresis. Reads MINDSHIFT_HEAT_HOLD_S once,
        #: here, so a test can set the env and construct a policy to pin either
        #: behaviour.
        self.hold = LoudnessHold(hold_s)
        # Ordered, de-duplicated: emission order is the caller's order, and a
        # repeated label would otherwise double-emit for that lane.
        self.channels: tuple[Channel, ...] = tuple(dict.fromkeys(channels))
        # Current level for each channel
        self._levels: dict[Channel, int] = {c: 0 for c in self.channels}
        # Last time we observed a qualifying event (≥ current level) for each channel
        self._last_qualifying_t: dict[Channel, float] = {c: 0.0 for c in self.channels}

    def on_events(
        self,
        events: list[VectorEvent],
        t: float,
        observed_s: float | None = WINDOW_S,
    ) -> list[NudgeEvent]:
        """Process vector events and return nudge events for level changes or de-escalation.

        Semantics: per channel, E = max scaled level from THIS call's events (0 if none).
        - If E > current: set level=E, emit, refresh qualifying time
        - Elif E == current and current > 0: refresh qualifying time, no emit (sustain clock)
        - Else (E < current): never snap down; if > cooldown_s since last qualifying event: drop exactly ONE level, emit, refresh time

        Events whose subscription routes to a channel this policy doesn't run
        are ignored (a single-channel caller subscribed to "hr_spike" simply
        never nudges on it) — they can't escalate a lane that doesn't exist.

        Args:
            events: List of VectorEvent objects to process
            t: Current timestamp
            observed_s: how many seconds of audio this call's observations
                cover. One 1 s window by default — the cadence the watch and
                the server's own PCM path tick on. A turn-driven caller
                (watch/relay.py) passes the turn's duration so the loudness
                hold counts seconds of speech, not calls.
                ``None`` means "this call carries NO loudness observation" and
                leaves the hold's run untouched. Required for every non-audio
                tick: an ``hr`` frame arriving mid-shout must not break a run
                the wearer's voice is still building, or the wrist would stay
                silent through a raised voice because a heart-rate sample
                landed between two windows.

        Returns:
            List of NudgeEvent objects, at most one per channel per call
        """
        nudges = []

        # Hold-N hysteresis, advanced exactly once per call that heard audio —
        # including quiet ones, since a window under the first rung is what
        # BREAKS a run. The raw, unscaled level is what clears the rung:
        # sensitivity is the wearer's preference about being told, not physics.
        if observed_s is None:
            loudness_may_escalate = True    # no loudness in this call to gate
        else:
            loud = any(e.vector == HOLD_VECTOR and e.level >= 1 for e in events)
            loudness_may_escalate = self.hold.observe(loud, observed_s)

        # Create a mapping of vector -> subscription for haptics-on subscriptions
        sub_by_vector: dict[str, VectorSubscription] = {}
        for sub in self.subs:
            if sub.haptics:  # Only consider haptics-on subscriptions
                sub_by_vector[sub.vector] = sub

        # Compute max scaled level from THIS call's events, grouped by channel
        event_max: dict[Channel, tuple[int, list[str]]] = {c: (0, []) for c in self.channels}
        for event in events:
            if event.vector == HOLD_VECTOR and not loudness_may_escalate:
                # Loud, but the rung has not held long enough yet. The call
                # then reads as E=0 for this lane, so it neither escalates nor
                # refreshes the sustain clock — the decay below runs exactly as
                # it would have on a quiet window.
                continue
            if event.vector in sub_by_vector:
                sub = sub_by_vector[event.vector]
                channel = sub.channel
                if channel not in event_max:
                    continue
                # Scale level by sensitivity using half-up rounding and clamp to 0-3
                scaled_level = self._round_half_up(event.level * sub.sensitivity)
                scaled_level = min(3, max(0, scaled_level))

                current_max, current_vecs = event_max[channel]
                if scaled_level > current_max:
                    event_max[channel] = (scaled_level, [event.vector])
                elif scaled_level == current_max and scaled_level > 0:
                    current_vecs.append(event.vector)

        # Process each channel
        for channel in self.channels:
            E, event_vectors = event_max[channel]
            current = self._levels[channel]

            if E > current:
                # Event(s) with higher level: escalate, emit, refresh clock
                self._levels[channel] = E
                self._last_qualifying_t[channel] = t
                nudges.append(NudgeEvent(
                    channel=channel,
                    level=E,
                    t=t,
                    vectors=sorted(set(event_vectors))
                ))
            elif E == current and current > 0:
                # Sustained event at current level: refresh clock, no emit
                self._last_qualifying_t[channel] = t
            else:
                # E < current or no events: check for de-escalation
                # Only drop if no qualifying event in more than cooldown_s seconds (strict >)
                if current > 0 and t - self._last_qualifying_t[channel] > self.cooldown_s:
                    # Drop exactly one level after cooldown with no high event
                    new_level = current - 1
                    self._levels[channel] = new_level
                    self._last_qualifying_t[channel] = t
                    nudges.append(NudgeEvent(
                        channel=channel,
                        level=new_level,
                        t=t,
                        vectors=[]
                    ))

        return nudges

    def current(self) -> dict[str, int]:
        """Return current level for each channel this policy runs."""
        return {c: self._levels[c] for c in self.channels}

    @staticmethod
    def _round_half_up(x: float) -> int:
        """Round using half-up method (match Kotlin Math.round), not banker's rounding.

        For x >= 0: int(x + 0.5)
        For x < 0: int(x - 0.5)
        This ensures 0.5 rounds up to 1, 1.5 to 2, etc., matching Kotlin Math.round (Task 10 mirror).
        """
        if x >= 0:
            return int(x + 0.5)
        else:
            return int(x - 0.5)

File: apps/mobile/src/live/nudgePolicy.ts
/**
 * Nudge/escalation policy — the TypeScript mirror of server/nudge_policy.py
 * (and the watch's NudgeStateMachine.kt). Thresholds/semantics must stay
 * identical across the three: per-channel hysteresis, strict-greater cooldown,
 * half-up rounding. The executable contract is
 * server/tests/fixtures/policy_vectors/nudge_policy.json, replayed here by
 * __tests__/liveNudgePolicy.test.ts for every case tagged `phone`/`server`.
 *
 * The phone runs a single lane ("A") over the coached user's OWN turns: a
 * level > 0 nudge fires a haptic (expo-haptics) and an on-screen flash. The
 * `selfTurnVectorEvents` helper turns a finalized self turn's prosody + text
 * tone into the vector events this policy consumes, using the SAME loudness
 * thresholds as the watch (+6/+10/+14 dB over baseline -> level 1/2/3).
 */

export type VectorName = "yelling" | "aggressive_tone" | "interrupting" | "airtime" | "hr_spike";
export type Channel = "A" | "B";

export interface VectorSubscription {
  vector: VectorName;
  sensitivity?: number;
  haptics?: boolean;
  channel?: Channel;
}

export interface VectorEvent {
  vector: VectorName;
  level: number;
  t: number;
  value?: number;
}

export interface NudgeEvent {
  channel: Channel;
  level: number;
  t: number;
  vectors: string[];
}

export const DEFAULT_CHANNELS: Channel[] = ["A", "B"];

/** Half-up rounding (Kotlin Math.round), not banker's — 0.5 -> 1, 1.5 -> 2. */
export function roundHalfUp(x: number): number {
  return x >= 0 ? Math.floor(x + 0.5) : Math.ceil(x - 0.5);
}

/** watch/models.py's default: hr_spike rides channel B, everything else A. */
function defaultChannel(vector: VectorName): Channel {
  return vector === "hr_spike" ? "B" : "A";
}

export class NudgePolicy {
  readonly channels: Channel[];
  private readonly subs: Required<VectorSubscription>[];
  private readonly levels = new Map<Channel, number>();
  private readonly lastQualifyingT = new Map<Channel, number>();

  constructor(
    subs: VectorSubscription[],
    private readonly cooldownS = 20.0,
    channels: Channel[] = DEFAULT_CHANNELS,
  ) {
    if (channels.length === 0) throw new Error("NudgePolicy needs at least one channel");
    this.channels = [...new Set(channels)];
    this.subs = subs.map((s) => ({
      vector: s.vector,
      sensitivity: s.sensitivity ?? 1.0,
      haptics: s.haptics ?? true,
      channel: s.channel ?? defaultChannel(s.vector),
    }));
    for (const c of this.channels) {
      this.levels.set(c, 0);
      this.lastQualifyingT.set(c, 0);
    }
  }

  onEvents(events: VectorEvent[], t: number): NudgeEvent[] {
    const nudges: NudgeEvent[] = [];
    const subByVector = new Map<string, Required<VectorSubscription>>();
    for (const s of this.subs) if (s.haptics) subByVector.set(s.vector, s);

    const eventMax = new Map<Channel, { level: number; vectors: string[] }>();
    for (const c of this.channels) eventMax.set(c, { level: 0, vectors: [] });
    for (const e of events) {
      const sub = subByVector.get(e.vector);
      if (!sub) continue;
      const slot = eventMax.get(sub.channel);
      if (!slot) continue; // a lane this policy doesn't run
      let scaled = roundHalfUp(e.level * sub.sensitivity);
      scaled = Math.min(3, Math.max(0, scaled));
      if (scaled > slot.level) {
        slot.level = scaled;
        slot.vectors = [e.vector];
      } else if (scaled === slot.level && scaled > 0) {
        slot.vectors.push(e.vector);
      }
    }

    for (const channel of this.channels) {
      const { level: E, vectors } = eventMax.get(channel)!;
      const current = this.levels.get(channel)!;
      if (E > current) {
        this.levels.set(channel, E);
        this.lastQualifyingT.set(channel, t);
        nudges.push({ channel, level: E, t, vectors: [...new Set(vectors)].sort() });
      } else if (E === current && current > 0) {
        this.lastQualifyingT.set(channel, t);
      } else if (current > 0 && t - this.lastQualifyingT.get(channel)! > this.cooldownS) {
        const next = current - 1;
        this.levels.set(channel, next);
        this.lastQualifyingT.set(channel, t);
        nudges.push({ channel, level: next, t, vectors: [] });
      }
    }
    return nudges;
  }

  current(): Record<string, number> {
    const out: Record<string, number> = {};
    for (const c of this.channels) out[c] = this.levels.get(c)!;
    return out;
  }
}

// ---------------------------------------------------------------------------
// Phone-side inputs: a finalized self turn -> vector events
// ---------------------------------------------------------------------------

/** The watch's loudness ladder (watch/vectors.py YELLING_LEVELS). */
export const YELLING_LEVELS: [number, number][] = [
  [14.0, 3],
  [10.0, 2],
  [6.0, 1],
];

export function yellingLevel(dbOverBaseline: number): number {
  for (const [threshold, level] of YELLING_LEVELS) {
    if (dbOverBaseline >= threshold) return level;
  }
  return 0;
}

/** Text-tone ladder for the aggressive_tone vector: the higher of the turn's
 *  frustration / defensiveness scores (0–100) -> 0..3. */
export function aggressiveToneLevel(frustration: number | null, defensiveness: number | null): number {
  const v = Math.max(frustration ?? 0, defensiveness ?? 0);
  if (v >= 85) return 3;
  if (v >= 70) return 2;
  if (v >= 50) return 1;
  return 0;
}

/**
 * Running loudness baseline over the user's own turns (median of what we've
 * heard so far), so "yelling" means "louder than YOU usually are on this
 * phone" — the same relative reasoning as the server's prosody labels.
 */
export class LoudnessBaseline {
  private readonly values: number[] = [];
  constructor(private readonly minTurns = 2) {}
  /** Baseline dBFS, or null until enough turns have been heard. */
  get value(): number | null {
    if (this.values.length < this.minTurns) return null;
    const s = [...this.values].sort((a, b) => a - b);
    const mid = s.length >> 1;
    return s.length % 2 ? s[mid] : (s[mid - 1] + s[mid]) / 2;
  }
  /** dB over the baseline for this turn (0 before a baseline exists), then
   *  fold the turn in. */
  observe(dbfs: number | null): number {
    if (dbfs === null || !Number.isFinite(dbfs)) return 0;
    const base = this.value;
    const over = base === null ? 0 : dbfs - base;
    this.values.push(dbfs);
    return over;
  }
}

export function selfTurnVectorEvents(
  t: number,
  dbOverBaseline: number,
  tone: { frustration: number | null; defensiveness: number | null },
): VectorEvent[] {
  return [
    { vector: "yelling", level: yellingLevel(dbOverBaseline), t, value: dbOverBaseline },
    {
      vector: "aggressive_tone",
      level: aggressiveToneLevel(tone.frustration, tone.defensiveness),
      t,
    },
  ];
}

/** The phone's single-lane default: both voice vectors on channel A. */
export function phoneNudgePolicy(cooldownS = 20.0): NudgePolicy {
  return new NudgePolicy(
    [
      { vector: "yelling", sensitivity: 1.0, haptics: true, channel: "A" },
      { vector: "aggressive_tone", sensitivity: 1.0, haptics: true, channel: "A" },
    ],
    cooldownS,
    ["A"],
  );
}

/** The haptic seam (expo-haptics in production, a spy in tests). */
export interface HapticSink {
  /** Level 1..3 -> light/medium/heavy; must never throw. */
  nudge(level: number): Promise<void>;
}
